python/lsst/ctrl/mpexec/cli/script/build.py

```diff
@@ -0,0 +1,113 @@
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from click import ClickException
+
+from ... import CmdLineFwk
+from ...cmdLineParser import (_ACTION_ADD_TASK,
+                              _ACTION_DELETE_TASK,
+                              _ACTION_CONFIG,
+                              _ACTION_CONFIG_FILE,
+                              _ACTION_ADD_INSTRUMENT)
+
+
+def build(config_file=[], config=[], delete=[], instrument=[], order_pipeline=None, pipeline_dot=None,
+          pipeline=None, save_pipeline=None, show=[], task=[]):
+    """Implements the command line interface `pipetask build` subcommand,
+    should only be called by command line tools and unit test code that tests
+    this function.
+
+    Build and optionally save pipeline definition.
+
+    Returns the pipeline instance that was built, for testing and for using
+    this function with other script functions.
+
+    Parameters
+    ----------
+    config_file : `list` [`str`]
+        Configuration override file(s), applies to a task with a given label.
+    config : `list` [`str`]
+        Configuration override(s) for a task with specified label. The format
+        of the string should be <label>:key=value, "task:foo=newfoo" or
+        "task:bar.baz=3".
+    delete : `list` [`str`]
+        Delete task with given label from pipeline.
+    instrument : `list` [`str`]
+        Add an instrument which will be used to load config overrides when
+        defining a pipeline. This must be the fully qualified class name.
+    order_pipeline : `bool`
+        If true, order tasks in pipeline based on their data dependencies,
+        ordering is performed as last step before saving or executing pipeline.
+    pipeline_dot : `str`
+        Path location for storing GraphViz DOT representation of a pipeline.
+    pipeline : `str`
+        Path location of a pipeline definition file in YAML format.
+    save_pipeline : `str`
+        Path location for storing resulting pipeline definition in YAML format.
+    show : `list` [`str`]
+        Descriptions of what to dump to stdout.
+    task : `list` [`str`]
+        Task name to add to pipeline, must be a fully qualified task name. Task
+        name can be followed by colon and label name, if label is not given
+        than task base name (class name) is used as label.
+
+    Returns
+    -------
+    pipeline : `lsst.pipe.base.Pipeline`
+        The pipeline instance that was built.
+    """
+
+    class Args:
```

Comment:
This is cute. We stuff random properties into an empty class? Couldn't we use a simple dataclass  here so that we can do some kind of checking?

Reply:
I was wanting to avoid managing the opaque `args` data structure much/any. Of course it's the idiom that comes from the argparse impl, IMHO it should be replaced entirely by member vars to the function or a smaller bunch of classes that contain related data members. I was hoping to avoid making that big of a change right now. But, doing one or the other is probably the Right Thing to do. I'll ponder, it might be worth a discussion on Slack... 🤔

Reply:
I didn't know any of the history (so this trick was used in the implementation by @andy-slac and you migrated it over?).  At the very least a comment like you just made in the class description for Args would be useful.

Reply:
the `args` object with a bunch of parser-defined properties is what you get from python's `argparse` module, e.g. the top example at https://docs.python.org/3/library/argparse.html. It's a common pattern (anti-pattern IMHO) to then pass around that object to your impl functions and let them pluck the members they need out of it. 
I can at least add a comment 👍